### PR TITLE
[Ray Serve] Add 2 examples of serving models on HPU

### PR DESCRIPTION
**Pull request by kira-lin**
## Why are these changes needed?

This PR adds 2 examples of using Ray Serve with HPU. One is for serving llama2-7b on single HPU, the other is for serving llama2-70b on 8 HPU using DeepSpeed.

In addition, it also enables "hccl" backend in air/util/torch_dist.py, which can be used to initialize torch dist of, for example, deepspeed workers.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/serve/doc_code/hpu_inference_serve.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        # enable hpu graph runtime
+        config["hpu_graphs"] = True
+        # lazy mode should be True when using hpu graphs
+        config["lazy_mode"] = True
+        # non streaming
+        if not streaming_response:
+            return self.generate(prompts, **config)
+        from starlette.responses import StreamingResponse
+        from transformers import TextIteratorStreamer
+        from functools import partial
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve.py:108]: The `starlette.responses` and `functools` imports can be moved to the top level of the doc, outside the deployments.

File: doc/source/serve/doc_code/hpu_inference_serve.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    print(outputs.text, flush=True)
+    sample_input["stream"] = True
+    input("Press Enter to start streaming inference")
+    outputs = requests.post("http://127.0.0.1:8000/", json=sample_input, stream=True)
+    outputs.raise_for_status()
+    for output in outputs.iter_content(chunk_size=None, decode_unicode=True):
+        print(output, end="", flush=True)
+    print()
+
+# __main_code_end__
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve.py:154]: Could you split the code example, so the entire Serve app is stored on one file, and the client script to query it is on another file? Ideally, the user should be able to run the serve app by running:

```
% ray start --head
% serve run your_example:app
```

and then separately query it using

```
python client.py
```

This makes it simple to productionize the example since the user can build a Serve config file directly from the example.

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+import ray
+import torch
+from transformers import TextStreamer
+from transformers import AutoTokenizer, AutoModelForCausalLM, AutoConfig
+from starlette.responses import StreamingResponse
+import deepspeed
+from optimum.habana.checkpoint_utils import (
+    get_ds_injection_policy,
+    write_checkpoints_json,
+)
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:16]:
```suggestion
import asyncio
import tempfile
from starlette.responses import StreamingResponse

import torch
from transformers import TextStreamer
from transformers import AutoTokenizer, AutoModelForCausalLM, AutoConfig

import ray
from ray.util.queue import Queue
from ray.runtime_env import RuntimeEnv

from ray import serve

import deepspeed
from optimum.habana.checkpoint_utils import (
    get_ds_injection_policy,
    write_checkpoints_json,
)
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+# habana required variables
+HABANA_ENVS = {
+    "PT_HPU_LAZY_ACC_PAR_MODE": "0",
+    "PT_HPU_ENABLE_REFINE_DYNAMIC_SHAPES": "0",
+    "PT_HPU_ENABLE_WEIGHT_CPU_PERMUTE": "0",
+    "PT_HPU_ENABLE_LAZY_COLLECTIVES": "true",
+    "HABANA_VISIBLE_MODULES": "0,1,2,3,4,5,6,7",
+}
+
+# set PyTorch distributed related environmental variables
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:29]: The code's comments sometimes refer to torch as `PyTorch` and sometimes as `torch`. Could you ensure that only one term is used?

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+from starlette.responses import StreamingResponse
+import deepspeed
+from optimum.habana.checkpoint_utils import (
+    get_ds_injection_policy,
+    write_checkpoints_json,
+)
+
+# the default port used by torch
+TORCH_DISTRIBUTED_DEFAULT_PORT = 29500
+# habana required variables
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:20]:
```suggestion
# variables required for habana
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+# __worker_def_start__
+@ray.remote(num_cpus=8, resources={"HPU": 1})
+class DeepSpeedInferenceWorker:
+    def __init__(self, model_id_or_path, world_size, local_rank):
+        # tweak transformers for better performance on gaudi
+        from optimum.habana.transformers.modeling_utils import (
+            adapt_transformers_to_gaudi,
+        )
+
+        adapt_transformers_to_gaudi()
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:81]: Please add a comment explaining what this function does.

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        else:
+
+            class DummyStreamer:
+                def put(self, value):
+                    pass
+
+                def end(self):
+                    pass
+
+            return DummyStreamer()
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:147]:
```suggestion
            class FakeStreamer:
                def put(self, value):
                    pass

                def end(self):
                    pass

            return FakeStreamer()
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        gen_tokens = self.model.generate(input_ids, **config)
+        return self.tokenizer.batch_decode(gen_tokens, skip_special_tokens=True)[0]
+
+    def streaming_generate(self, prompt, streamer, **config):
+        input_ids = self.tokenizer(
+            prompt, return_tensors="pt", padding=True
+        ).input_ids.to(self.device)
+        self.model.generate(input_ids, streamer=streamer, **config)
+
+    def get_streamer(self):
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:135]: Could you add a comment explaining why we need different streamers here?

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+        for i in range(world_size):
+            worker_env = {**HABANA_ENVS, **TORCH_ENVS}
+            worker_env["RANK"] = str(i)
+            worker_env["LOCAL_RANK"] = str(i)
+            self.deepspeed_workers.append(
+                DeepSpeedInferenceWorker.options(
+                    runtime_env=RuntimeEnv(env_vars=worker_env)
+                ).remote(model_id_or_path, world_size, i)
+            )
+        # get workers' streamer
```
Comment (shrekris-anyscale) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:169]:
```suggestion
        # get workers' streamers
```

**Review by shrekris-anyscale — COMMENTED** [review]
Thanks for contributing this @kira-lin! These docs will be helpful for users trying to serve their models on Gaudi.

A couple high level comments:

1. Some of the Gaudi and HuggingFace-specific syntax makes the code a bit hard to follow (e.g. the `bos_token_id`, `eos_token_id`, the `config` that's passed in through the request, etc). Is there any way we can remove or simplify some of these settings?
2. Mixing the client logic into the same file as the one running the Serve app makes it tough to differentiate the client and server logic. Can we split them into separate scripts?

Please let me know if you have any questions about my other comments.

**Comment by kira-lin** [comment]
Hi @shrekris-anyscale ,
I tried to split them to serve part and client part. I also simplified things like `pad_token_id` a bit. Please have a review again, thanks

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+* You want to serve a large language model on HPU
+
+* You want to serve a large language model using DeepSpeed on HPU
+
+## Environment Setup
+
+We recommend using a prebuilt container to run these examples. To run a container, you will need Docker. Please refer to [here](https://docs.docker.com/engine/install/) to install it if you haven't.
+
+Next, ensure that the drivers and container runtime
+are installed, if not, refer to [this guide](https://docs.habana.ai/en/latest/Installation_Guide/Bare_Metal_Fresh_OS.html?highlight=installer#run-using-containers). To verify your installation, start a shell and run `hl-smi`, and it should print information like this:
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:34]:
```suggestion
Next, follow [this guide](https://docs.habana.ai/en/latest/Installation_Guide/Bare_Metal_Fresh_OS.html?highlight=installer#run-using-containers) to install the Habana drivers and container runtime. To verify your installation, start a shell and run `hl-smi`. It should print status information about the HPUs on the machine:
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+|   2        N/A   N/A    N/A                                      N/A        |
+|   3        N/A   N/A    N/A                                      N/A        |
+|   4        N/A   N/A    N/A                                      N/A        |
+|   5        N/A   N/A    N/A                                      N/A        |
+|   6        N/A   N/A    N/A                                      N/A        |
+|   7        N/A   N/A    N/A                                      N/A        |
++=============================================================================+
+```
+
+If it works, you can then start a container like this:
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:82]:
```suggestion
Next, start the Habana container:
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+```bash
+docker pull vault.habana.ai/gaudi-docker/1.14.0/ubuntu22.04/habanalabs/pytorch-installer-2.1.1:latest
+docker run -it --runtime=habana -e HABANA_VISIBLE_DEVICES=all -e OMPI_MCA_btl_vader_single_copy_mechanism=none --cap-add=sys_nice --net=host --ipc=host vault.habana.ai/gaudi-docker/1.14.0/ubuntu22.04/habanalabs/pytorch-installer-2.1.1:latest
+```
+
+Note that you may want to mount the directory containing the examples into the container, as well as the models.
+
+Inside the container, run
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:90]:
```suggestion
To follow the examples in this tutorial, mount the directory containing the examples and models into the container. Inside the container, run
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+```bash
+pip install ray[tune,serve]
+pip install git+https://github.com/huggingface/optimum-habana.git
+# 1.14.0 should be replaced with the driver version of the container
+pip install git+https://github.com/HabanaAI/DeepSpeed.git@1.14.0
+# Only needed by the deepspeed example
+export RAY_EXPERIMENTAL_NOSET_HABANA_VISIBLE_MODULES=1
+```
+
+Start a ray cluster via `ray start --head` and you are ready to run the examples.
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:100]: Should they run this inside the container? If so could you replace this with:

```suggestion
Start a Ray in the container via `ray start --head`. You are now ready to run the examples.
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+
+## Single HPU Example
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:102]:
```suggestion
## Running a model on a single HPU
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+
+This example shows how to deploy a Llama2-7b model on Gaudi for inference.
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:104]: Try to use one term (HPU/Gaudi) consistently; otherwise, it get a bit confusing what the tutorial is referring to.

```suggestion
This example shows how to deploy a Llama2-7b model on an HPU for inference. 
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+export RAY_EXPERIMENTAL_NOSET_HABANA_VISIBLE_MODULES=1
+```
+
+Start a ray cluster via `ray start --head` and you are ready to run the examples.
+
+## Single HPU Example
+
+This example shows how to deploy a Llama2-7b model on Gaudi for inference. 
+
+First, define a deployment that serves Llama2-7b model using a HPU. Note that we enable [HPU graph optimization](https://docs.habana.ai/en/latest/Gaudi_Overview/SynapseAI_Software_Suite.html?highlight=graph#graph-compiler-and-runtime) for better performance.
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:106]:
```suggestion
First, define a deployment that serves a Llama2-7b model using an HPU. Note that we enable [HPU graph optimizations](https://docs.habana.ai/en/latest/Gaudi_Overview/SynapseAI_Software_Suite.html?highlight=graph#graph-compiler-and-runtime) for better performance.
```

Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:106]: Could you double check that the tutorial uses _an_ HPU, not _a_ HPU?

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+:end-before: __model_def_end__
+```
+
+Copy the code above and save it as `hpu_inference_serve.py`. Start the deployment like this:
+
+```bash
+serve run hpu_inference_serve:entrypoint
+```
+
+If it succeeds, you should see something like this:
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:120]:
```suggestion
The terminal should print logs as the deployment starts up:
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+```text
+Once upon a time, in a far-off land, there was a magical kingdom called "Happily Ever Laughter." It was a place where laughter was the key to unlocking all the joys of life, and where everyone lived in perfect harmony.
+In this kingdom, there was a beautiful princess named Lily. She was kind, gentle, and had a heart full of laughter. Every day, she would wake up with a smile on her face, ready to face whatever adventures the day might bring.
+One day, a wicked sorcerer cast a spell on the kingdom, causing all
+in a far-off land, there was a magical kingdom called "Happily Ever Laughter." It was a place where laughter was the key to unlocking all the joys of life, and where everyone lived in perfect harmony.
+In this kingdom, there was a beautiful princess named Lily. She was kind, gentle, and had a heart full of laughter. Every day, she would wake up with a smile on her face, ready to face whatever adventures the day might bring.
+One day, a wicked sorcerer cast a spell on the kingdom, causing all
+```
+
+## Multiple HPU Example
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:179]:
```suggestion
## Running a sharded model on a multiple HPUs
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+
+This example shows how to deploy a Llama2-70b model using 8 HPUs orchestrated by DeepSpeed. 
+
+The example requires caching the Llama2-70b model. If you haven't cached it yet, run the following Python code in your container to cache the model.
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:183]:
```suggestion
The example requires caching the Llama2-70b model. Run the following Python code in the Habana container to cache the model. 
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+
+Next, define a deployment:
+
+```{literalinclude} ../doc_code/hpu_inference_serve_deepspeed.py
+:language: python
+:start-after: __deploy_def_start__
+:end-before: __deploy_def_end__
+```
+
+Copy both blocks of code above and save them into `hpu_inference_serve_deepspeed.py`. You can run this example using `serve run hpu_inference_serve_deepspeed:entrypoint`.
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:212]:
```suggestion
Copy both blocks of code above and save them into `hpu_inference_serve_deepspeed.py`. Run this example using `serve run hpu_inference_serve_deepspeed:entrypoint`.
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+
+If it succeds, you should see something like this:
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:214]:
```suggestion
The terminal should print logs as the deployment starts up:
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+(ServeController pid=44317) WARNING 2024-02-01 06:13:27,770 controller 44317 deployment_state.py:2171 - Deployment 'DeepSpeedLlamaModel' in application 'default' has 1 replicas that have taken more than 30s to initialize. This may be caused by a slow __init__ or reconfigure method.
+Loading 15 checkpoint shards:  87%|████████▋ | 13/15 [04:24<00:34, 17.26s/it] [repeated 8x across cluster]
+(ServeController pid=44317) WARNING 2024-02-01 06:13:57,873 controller 44317 deployment_state.py:2171 - Deployment 'DeepSpeedLlamaModel' in application 'default' has 1 replicas that have taken more than 30s to initialize. This may be caused by a slow __init__ or reconfigure method.
+Loading 15 checkpoint shards:  87%|████████▋ | 13/15 [04:25<00:34, 17.35s/it] [repeated 7x across cluster]
+Loading 15 checkpoint shards:  93%|█████████▎| 14/15 [04:44<00:17, 17.55s/it]
+Loading 15 checkpoint shards: 100%|██████████| 15/15 [05:02<00:00, 18.30s/it] [repeated 8x across cluster]
+2024-02-01 06:14:24,054 SUCC scripts.py:483 -- Deployed Serve app successfully.
+```
+
+Use the same code snippet introduced in Single HPU Example for sending requests to perform generation. An example output:
```
Comment (shrekris-anyscale) [inline doc/source/serve/tutorials/hpu_inference_serve.md:283]:
```suggestion
Use the same code snippet introduced in Single HPU Example to send generation requests. Here's an example output:
```

**Review by shrekris-anyscale — COMMENTED** [review]
Really nice work! The code is much more readable, and the docs are clearer. I added some more comments, but they're largely minor formatting corrections. Otherwise, the tutorial looks good to me.

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 23]
```diff
@@ -0,0 +1,262 @@
...
+        if value == self.stop_signal:
+            raise StopIteration()
+        else:
+            return value
+
+
+# __worker_def_end__
+
+# __deploy_def_start__
+# variables required for DeepSpeed on HPU
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:156]:
```suggestion
# Required variables for DeepSpeed on HPU
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 24]
```diff
@@ -0,0 +1,262 @@
...
+HABANA_ENVS = {
+    "PT_HPU_LAZY_ACC_PAR_MODE": "0",
+    "PT_HPU_ENABLE_REFINE_DYNAMIC_SHAPES": "0",
+    "PT_HPU_ENABLE_WEIGHT_CPU_PERMUTE": "0",
+    "PT_HPU_ENABLE_LAZY_COLLECTIVES": "true",
+    "HABANA_VISIBLE_MODULES": "0,1,2,3,4,5,6,7",
+}
+
+
+# Define the Ray serve deployment
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:166]:
```suggestion
# Define the Ray Serve deployment
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+        for worker, streamer in zip(self.deepspeed_workers, self.streamers):
+            worker.streaming_generate.remote(prompt, streamer, **config)
+
+    def consume_streamer(self, streamer):
+        """Consume the streamer, return a generator"""
+        for token in streamer:
+            yield token
+
+    async def __call__(self, http_request):
+        """Handle received http requests"""
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:221]:
```suggestion
        """Handle received HTTP requests."""
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 26]
```diff
@@ -0,0 +1,262 @@
...
+
+        # Load fields from the request
+        json_request: str = await http_request.json()
+        text = json_request["text"]
+        # Config used in generation
+        config = json_request["config"] if "config" in json_request else {}
+        streaming_response = json_request["stream"]
+
+        # prepare prompts
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:230]:
```suggestion
        # Prepare prompts
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 27]
```diff
@@ -0,0 +1,262 @@
...
+        prompts = []
+        if isinstance(text, list):
+            prompts.extend(text)
+        else:
+            prompts.append(text)
+
+        # process config
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:237]:
```suggestion
        # Process config
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 28]
```diff
@@ -0,0 +1,262 @@
...
+        if "max_new_tokens" not in config:
+            # if not specified, use default value
+            config["max_new_tokens"] = 128
+
+        # enable hpu graph runtime
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:242]:
```suggestion
        # Enable HPU graph runtime
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 29]
```diff
@@ -0,0 +1,262 @@
...
+        config["hpu_graphs"] = True
+        # lazy mode should be True when using hpu graphs
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:244]:
```suggestion
        # Lazy mode should be True when using HPU graphs
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 30]
```diff
@@ -0,0 +1,262 @@
...
+        config["lazy_mode"] = True
+
+        # non streaming case
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:247]:
```suggestion
        # Non-streaming case
```

File: doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py [hunk 31]
```diff
@@ -0,0 +1,262 @@
...
+        # streming case
+        self.streaming_generate(prompts, **config)
+        return StreamingResponse(
+            self.consume_streamer(self.streamers[0]),
+            status_code=200,
+            media_type="text/plain",
+        )
+
+
+# replace the model id with path if necessary
```
Comment (angelinalg) [inline doc/source/serve/doc_code/hpu_inference_serve_deepspeed.py:260]:
```suggestion
# Replace the model ID with path if necessary
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 32]
```diff
@@ -0,0 +1,291 @@
+# Serving an inference model on Intel Habana Gaudi
+
+[Habana® Gaudi® AI Processor (HPU)](https://habana.ai) training processors are AI hardware accelerators designed by Habana Labs. They are designed to maximize training throughput and efficiency, while providing developers with optimized software and tools that scale to many workloads and systems. For more information, check out [Gaudi Architecture](https://docs.habana.ai/en/latest/Gaudi_Overview/index.html) and [Gaudi Developer Docs](https://developer.habana.ai/).
+
+In this tutorial, we will go through two examples:
```
Comment (angelinalg) [inline doc/source/serve/tutorials/hpu_inference_serve.md:5]:
```suggestion
This tutorial has two examples:
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 33]
```diff
@@ -0,0 +1,291 @@
...
+
+    * How to stream response from DeepSpeed workers
+
+This tutorial should help you with following use cases:
+
+* You want to serve a large language model on HPU
+
+* You want to serve a large language model using DeepSpeed on HPU
+
+## Environment Setup
```
Comment (angelinalg) [inline doc/source/serve/tutorials/hpu_inference_serve.md:29]:
```suggestion
## Environment setup
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 34]
```diff
@@ -0,0 +1,291 @@
...
+docker run -it --runtime=habana -e HABANA_VISIBLE_DEVICES=all -e OMPI_MCA_btl_vader_single_copy_mechanism=none --cap-add=sys_nice --net=host --ipc=host vault.habana.ai/gaudi-docker/1.14.0/ubuntu22.04/habanalabs/pytorch-installer-2.1.1:latest
+```
+
+Note that you may want to mount the directory containing the examples into the container, as well as the models.
+
+Inside the container, run
+```bash
+pip install ray[tune,serve]
+pip install git+https://github.com/huggingface/optimum-habana.git
+# 1.14.0 should be replaced with the driver version of the container
```
Comment (angelinalg) [inline doc/source/serve/tutorials/hpu_inference_serve.md:94]:
```suggestion
# Replace 1.14.0 with the driver version of the container
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 35]
```diff
@@ -0,0 +1,291 @@
...
+pip install git+https://github.com/HabanaAI/DeepSpeed.git@1.14.0
+# Only needed by the deepspeed example
```
Comment (angelinalg) [inline doc/source/serve/tutorials/hpu_inference_serve.md:96]:
```suggestion
# Only needed by the DeepSpeed example
```

File: doc/source/serve/tutorials/hpu_inference_serve.md (hunk 16 above)
Comment (angelinalg) [reply doc/source/serve/tutorials/hpu_inference_serve.md:106]: From a tech writing perspective, it should be _an_ HPU, based on the way you would pronounce the letter "H". https://developers.google.com/style/articles

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 36]
```diff
@@ -0,0 +1,291 @@
...
+
+This example shows how to deploy a Llama2-70b model using 8 HPUs orchestrated by DeepSpeed. 
+
+The example requires caching the Llama2-70b model. If you haven't cached it yet, run the following Python code in your container to cache the model. 
+
+```python
+from huggingface_hub import snapshot_download
+snapshot_download(
+    "meta-llama/Llama-2-70b-chat-hf",
+    # replace the path if necessary
```
Comment (angelinalg) [inline doc/source/serve/tutorials/hpu_inference_serve.md:189]:
```suggestion
    # Replace the path if necessary
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 37]
```diff
@@ -0,0 +1,291 @@
...
+    cache_dir=os.getenv("TRANSFORMERS_CACHE", None),
+    # specify your huggingface token
```
Comment (angelinalg) [inline doc/source/serve/tutorials/hpu_inference_serve.md:191]:
```suggestion
    # Specify your Hugging Face token
```

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 38]
```diff
@@ -0,0 +1,291 @@
...
+    token=""
+)
+```
+
+In this example, the deployment replica sends prompts to the deepspeed workers, which are running in ray actors:
```
Comment (angelinalg) [inline doc/source/serve/tutorials/hpu_inference_serve.md:196]:
```suggestion
In this example, the deployment replica sends prompts to the Deepspeed workers, which are running in Ray actors:
```

**Review by angelinalg — APPROVED** [review]
Just some nits. Thanks for putting this together!

**Review by jerome-habana — APPROVED** [review]

**Comment by shrekris-anyscale** [comment]
@kira-lin thanks again for submitting this contribution! It looks nearly ready to merge.

One pending issue is the code changes to the private APIs in the `python/ray/air/util/torch_dist.py` file. I synced with the team, and they’re actually planning to remove this file shortly. We don’t currently have the bandwidth to maintain the file and provide support for the DeepSpeed example.

Could we do the following:

1. Update this PR to remove the `Running a sharded model on multiple HPUs` section.
2. Remove the code changes to `python/ray/air/util/torch_dist.py`.
3. If feasible, host the documentation for `Running a sharded model on multiple HPUs` in the Habana docs?

We can merge this PR once the first two changes are made. In the future, we can also add a link in the Ray docs to the Habana docs for `Running a sharded model on multiple HPUs`.

**Comment by jerome-habana** [comment]
> @kira-lin thanks again for submitting this contribution! It looks nearly ready to merge.
> 
> One pending issue is the code changes to the private APIs in the `python/ray/air/util/torch_dist.py` file. I synced with the team, and they’re actually planning to remove this file shortly. We don’t currently have the bandwidth to maintain the file and provide support for the DeepSpeed example.
> 
> Could we do the following:
> 
> 1. Update this PR to remove the `Running a sharded model on multiple HPUs` section.
> 2. Remove the code changes to `python/ray/air/util/torch_dist.py`.
> 3. If feasible, host the documentation for `Running a sharded model on multiple HPUs` in the Habana docs?
> 
> We can merge this PR once the first two changes are made. In the future, we can also add a link in the Ray docs to the Habana docs for `Running a sharded model on multiple HPUs`.

@shrekris-anyscale  this change is needed for multi-card runs.

**Comment by shrekris-anyscale** [comment]
> this change is needed for multi-card runs.

To clarify, are multi-card runs the same as running with multiple HPUs? Is running on a single HPU still possible without the change?

**Comment by kira-lin** [comment]
> To clarify, are multi-card runs the same as running with multiple HPUs?

Yes, multi-card refers to multiple HPUs.

> Is running on a single HPU still possible without the change?

Yes, running on a single HPU does not require this change. However, torch_dist.py is handy when initializing torch dist among inference workers. Our open-source inference framework, [llm-on-ray](https://github.com/intel/llm-on-ray) also utilizes this file in its deepspeed inference workers. Are there any alternatives for the same function in Ray?

File: doc/source/serve/tutorials/hpu_inference_serve.md [hunk 39]
```diff
@@ -0,0 +1,284 @@
...
+```text
+Once upon a time, in a far-off land, there was a magical kingdom called "Happily Ever Laughter." It was a place where laughter was the key to unlocking all the joys of life, and where everyone lived in perfect harmony.
+In this kingdom, there was a beautiful princess named Lily. She was kind, gentle, and had a heart full of laughter. Every day, she would wake up with a smile on her face, ready to face whatever adventures the day might bring.
+One day, a wicked sorcerer cast a spell on the kingdom, causing all
+in a far-off land, there was a magical kingdom called "Happily Ever Laughter." It was a place where laughter was the key to unlocking all the joys of life, and where everyone lived in perfect harmony.
+In this kingdom, there was a beautiful princess named Lily. She was kind, gentle, and had a heart full of laughter. Every day, she would wake up with a smile on her face, ready to face whatever adventures the day might bring.
+One day, a wicked sorcerer cast a spell on the kingdom, causing all
+```
+
+## Running a sharded model on multiple HPUs
```
[the remark this answers is not in the document]
Comment (kira-lin) [reply doc/source/serve/tutorials/hpu_inference_serve.md]: resolved

**Comment by zcin** [comment]
@kira-lin The docs build is failing because the new doc isn't added to any toctree. Could you add it to the toctree [here](https://docs.ray.io/en/latest/serve/tutorials/index.html), near [this tutorial](https://docs.ray.io/en/latest/serve/tutorials/aws-neuron-core-inference.html)?

**Comment by kira-lin** [comment]
fixed @zcin

**Review by zcin — APPROVED** [review]
